src/coreclr/jit/liveness.cpp

```diff
@@ -376,6 +363,63 @@ void Compiler::fgPerNodeLocalVarLiveness(GenTree* tree)
     }
 }
 
+#if defined(FEATURE_HW_INTRINSICS)
+void Compiler::fgPerNodeLocalVarLiveness(GenTreeHWIntrinsic* hwintrinsic)
+{
+    NamedIntrinsic intrinsicId = hwintrinsic->GetHWIntrinsicId();
+
+    // We can't call fgMutateGcHeap unless the block has recorded a MemoryDef
+    //
+    if (hwintrinsic->OperIsMemoryStore())
+    {
+        // We currently handle this like a Volatile store, so it counts as a definition of GcHeap/ByrefExposed
+        fgCurMemoryDef |= memoryKindSet(GcHeap, ByrefExposed);
+    }
+    else if (hwintrinsic->OperIsMemoryLoad())
+    {
+        // This instruction loads from memory and we need to record this information
+        fgCurMemoryUse |= memoryKindSet(GcHeap, ByrefExposed);
+    }
+#if defined(TARGET_XARCH)
+    else if (HWIntrinsicInfo::HasSpecialSideEffect(intrinsicId))
+    {
+        switch (intrinsicId)
+        {
+            case NI_SSE_StoreFence:
+            case NI_SSE2_LoadFence:
+            case NI_SSE2_MemoryFence:
+            case NI_X86Serialize_Serialize:
+            {
+                // While these don't technically do an assignment, they are modeled the same as
+                // GT_MEMORYBARRIER which tracks itself as a definition of GcHeap/ByrefExposed
+
+                fgCurMemoryDef |= memoryKindSet(GcHeap, ByrefExposed);
+                break;
+            }
+
+            case NI_SSE_Prefetch0:
+            case NI_SSE_Prefetch1:
+            case NI_SSE_Prefetch2:
+            case NI_SSE_PrefetchNonTemporal:
+            {
+                // These instructions don't technically load from memory, but they do take and
+                // consume an address in a non-faulting way, so we still want to record this as
+                // a memory use.
+
+                fgCurMemoryUse |= memoryKindSet(GcHeap, ByrefExposed);
+                break;
```

Comment:
Why is this needed?

In principle, liveness only cares about actual uses of memory, i. e. ones where VN reads the "current memory VN" heap.

Reply:
These are effectively "reading" the address and will load that data into the L1 cache.

It's just that it doesn't also fault like a regular load would, even if the address is invalid.

Reply:
In that case I don't think this code is needed -- as I said, memory liveness only cares about cases that are handled in VN as loads. E. g. it doesn't need to count tracked locals as memory uses (even as they may not get a register).

(Concretely, what this does is ensure memory PHIs get added to blocks that "use" memory)

Reply:
This is actually consuming a pointer to arbitrary memory however, not a local?

Reply:
Memory liveness "use" != physical memory "use".

Concretely, each case where we add a "use" here corresponds to some code in VN which references `fgCurMemoryVN`. Thus, since these intrinsics do not need special numbering dependent on `fgCurMemoryVN`, they don't need to be handled here - it's just dead code.

Reply:
Wouldn't this need to be considered live though in the case its a byref for example?

Reply:
The address itself will be considered live (by virtue of being used by an operator), memory liveness is orthogonal to that.

Reply:
Removed.